In [1]:
from pathlib import Path
import numpy as np
import sys
import time

from mfa_processor import MFA_PROCESSOR

True
True


In [2]:
DIR_MFA = Path.cwd().parent.parent / 'MFA_code'
print(DIR_MFA.exists())

sys.path.append(str(DIR_MFA))
from mfa import MFA

True


# Constants

In [3]:
PATH_DATA = Path.cwd().parent.parent.parent / 'data'
PATH_DATA.exists()

True

In [4]:
PATH_ARCHAEA_FNA = (PATH_DATA / "raw" / "genomes" / "archaea" /
    "genomic_fna")
PATH_ARCHAEA_FNA.exists()

True

# Processing

In [5]:
mfa_processor = MFA_PROCESSOR(PATH_ARCHAEA_FNA)

## Decompression

In [6]:
# show the elements inside the folder PATH_ARCHAEA_FNA
# in alphabetical order

# elements in the folder PATH_ARCHAEA_FNA
elements = list(PATH_ARCHAEA_FNA.iterdir())
elements = sorted(elements)
print(len(elements))

test_list_1 = []
for element in elements[:3]:
    test_list_1.append(element.name)

for element in test_list_1:
    print(element)


1193
acidianus_ambivalens
acidianus_brierleyi
acidianus_hospitalis


In [7]:
dirs_to_iterate = test_list_1  # List of subdirectory names to process
mfa_processor.decompress(dirs_to_iterate)

Decompressed: at acidianus_ambivalens to /home/diegoflm/Documents/UV/courses_and_subjects/final_year_proyect/project/data/raw/genomes/archaea/genomic_fna/acidianus_ambivalens/GCF_009428885.1_ASM942888v1_genomic.fna
Decompressed: at acidianus_brierleyi to /home/diegoflm/Documents/UV/courses_and_subjects/final_year_proyect/project/data/raw/genomes/archaea/genomic_fna/acidianus_brierleyi/GCF_003201835.1_ASM320183v1_genomic.fna
Decompressed: at acidianus_hospitalis to /home/diegoflm/Documents/UV/courses_and_subjects/final_year_proyect/project/data/raw/genomes/archaea/genomic_fna/acidianus_hospitalis/GCF_000213215.1_ASM21321v1_genomic.fna


## get_path_fna()

In [8]:
elements[:3]

[PosixPath('/home/diegoflm/Documents/UV/courses_and_subjects/final_year_proyect/project/data/raw/genomes/archaea/genomic_fna/acidianus_ambivalens'),
 PosixPath('/home/diegoflm/Documents/UV/courses_and_subjects/final_year_proyect/project/data/raw/genomes/archaea/genomic_fna/acidianus_brierleyi'),
 PosixPath('/home/diegoflm/Documents/UV/courses_and_subjects/final_year_proyect/project/data/raw/genomes/archaea/genomic_fna/acidianus_hospitalis')]

In [9]:
fna_paths = []

for organism in elements[:3]:
    fna_paths.append(mfa_processor.get_path_fna(organism))
    print(fna_paths[-1])

/home/diegoflm/Documents/UV/courses_and_subjects/final_year_proyect/project/data/raw/genomes/archaea/genomic_fna/acidianus_ambivalens/GCF_009428885.1_ASM942888v1_genomic.fna
/home/diegoflm/Documents/UV/courses_and_subjects/final_year_proyect/project/data/raw/genomes/archaea/genomic_fna/acidianus_brierleyi/GCF_003201835.1_ASM320183v1_genomic.fna
/home/diegoflm/Documents/UV/courses_and_subjects/final_year_proyect/project/data/raw/genomes/archaea/genomic_fna/acidianus_hospitalis/GCF_000213215.1_ASM21321v1_genomic.fna


## extract_sequence()
Sequence Extraction

### dev

In [10]:
elements = list(PATH_ARCHAEA_FNA.iterdir())
elements = sorted(elements)
print(len(elements))

1193


In [11]:
elements[0]

PosixPath('/home/diegoflm/Documents/UV/courses_and_subjects/final_year_proyect/project/data/raw/genomes/archaea/genomic_fna/acidianus_ambivalens')

In [12]:
organism_file_names = list(elements[0].iterdir())
organism_file_names

fna_files = []
for name in organism_file_names:
    if str(name).endswith('.fna'):    
        fna_files.append(name)

print(fna_files)
assert(len(fna_files) == 1)

path_organism_genome = fna_files[0]
path_organism_genome

[PosixPath('/home/diegoflm/Documents/UV/courses_and_subjects/final_year_proyect/project/data/raw/genomes/archaea/genomic_fna/acidianus_ambivalens/GCF_009428885.1_ASM942888v1_genomic.fna')]


PosixPath('/home/diegoflm/Documents/UV/courses_and_subjects/final_year_proyect/project/data/raw/genomes/archaea/genomic_fna/acidianus_ambivalens/GCF_009428885.1_ASM942888v1_genomic.fna')

In [13]:
with open(path_organism_genome, 'r') as f:
            fna_content = f.read()
            f.close()
seq_segments = fna_content.split('\n')
print(f"len(seq_segments): {len(seq_segments)}")
seq = ""
# for genome_line in seq_segments[:]:
#     if (not genome_line.startswith('>')) and ( len(genome_line) > 0 ):
#         seq += genome_line

for genome_line in seq_segments[:]:
    if genome_line.startswith('>') or (len(genome_line) == 0):
        continue
    else:
        seq += genome_line
        

# for genome_line in seq:
#     print(genome_line)

print(len(seq))
print(f"last line length: {len(seq_segments[-1])}" )
print(f"last line: {(seq_segments[-1])}" )
print(f"last character: {seq[-1]}")
print(type(seq[-1]))
print(seq[-15:])

print("Out place characters:")
for char in seq:
    if char not in ['A', 'C', 'G', 'T']:
        print(char)

len(seq_segments): 29185
2326940
last line length: 0
last line: 
last character: C
<class 'str'>
ATGGACAGCGTGAAC
Out place characters:


In [14]:
# len(seq_segments): 29185
# 2326940
# last line length: 1
# last line: C
# <class 'str'>
# ATGGACAGCGTGAAC
# Out place characters:

### using mfa_processor

In [15]:
fna_paths[0] == path_organism_genome

True

In [16]:
sequence_from_method, metadata = mfa_processor.extract_sequence(path_organism_genome)

print(len(sequence_from_method))
print(f"last character: {sequence_from_method[-1]}")
print(type(sequence_from_method[-1]))
print(sequence_from_method[-15:])

print("Out place characters:")
for char in sequence_from_method:
    if char not in ['A', 'C', 'G', 'T']:
        print(char)

2326940
last character: C
<class 'str'>
ATGGACAGCGTGAAC
Out place characters:


len(seq_segments): 29185  
2326940  
last line length: 0  
last line:  
last character: C  
<class 'str'>  
ATGGACAGCGTGAAC  
Out place characters:  

In [17]:
sequence_from_method == seq

True

In [18]:
path_organism_genome

PosixPath('/home/diegoflm/Documents/UV/courses_and_subjects/final_year_proyect/project/data/raw/genomes/archaea/genomic_fna/acidianus_ambivalens/GCF_009428885.1_ASM942888v1_genomic.fna')

## compute_gc_mfa()
GC content and MFA

In [19]:
instance_mfa = MFA(sequence_from_method)

In [20]:
instance_mfa.gc_content()

0.3436646411166597

In [21]:
dir_acidianus_ambivalens = elements[0]
dir_acidianus_ambivalens

PosixPath('/home/diegoflm/Documents/UV/courses_and_subjects/final_year_proyect/project/data/raw/genomes/archaea/genomic_fna/acidianus_ambivalens')

In [22]:
directory_paths = elements[0:5]
directory_paths



[PosixPath('/home/diegoflm/Documents/UV/courses_and_subjects/final_year_proyect/project/data/raw/genomes/archaea/genomic_fna/acidianus_ambivalens'),
 PosixPath('/home/diegoflm/Documents/UV/courses_and_subjects/final_year_proyect/project/data/raw/genomes/archaea/genomic_fna/acidianus_brierleyi'),
 PosixPath('/home/diegoflm/Documents/UV/courses_and_subjects/final_year_proyect/project/data/raw/genomes/archaea/genomic_fna/acidianus_hospitalis'),
 PosixPath('/home/diegoflm/Documents/UV/courses_and_subjects/final_year_proyect/project/data/raw/genomes/archaea/genomic_fna/acidianus_infernus'),
 PosixPath('/home/diegoflm/Documents/UV/courses_and_subjects/final_year_proyect/project/data/raw/genomes/archaea/genomic_fna/acidianus_manzaensis')]

In [23]:
path_organism_genome

PosixPath('/home/diegoflm/Documents/UV/courses_and_subjects/final_year_proyect/project/data/raw/genomes/archaea/genomic_fna/acidianus_ambivalens/GCF_009428885.1_ASM942888v1_genomic.fna')

In [24]:
len(directory_paths)

5

In [25]:
# measure time
initial_time = time.time()
    
for dir in directory_paths[:1]:
    dir_files = dir.glob('*')
    fna_counter = 0
    seq = ""
    genome_file = None
    for file in dir_files:
        if file.name.endswith('.fna'):
            fna_counter += 1
            genome_file = file
            print(f"genome_file: {genome_file}")
    print(f"fna_counter: {fna_counter}")
    assert( fna_counter == 1 or fna_counter == 0 )
    time_stamp_1 = time.time()
    time_lapse_1 = time_stamp_1 - initial_time
    print(f"time_lapse_1: {time_lapse_1}")
    seq, metadata = mfa_processor.extract_sequence( genome_file )
    time_stamp_2 = time.time()
    time_lapse_2 = time_stamp_2 - time_stamp_1
    print(f"time_lapse_2: {time_lapse_2}")
    gc_content, Dq_vals, r_squared, Dq = mfa_processor.compute_gc_Dq(seq, use_powers = True, power = 13)
    time_stamp_3 = time.time()
    time_lapse_3 = time_stamp_3 - time_stamp_2
    print(f"time_lapse_3: {time_lapse_3}")
    
    print(f"gc_content: {gc_content}")
    print(f"Dq_vals: {Dq_vals}")
    print(f"r_squared: {r_squared}")
    print(f"Dq: {Dq}")

    if seq is not None:
        print(seq[-5:])

genome_file: /home/diegoflm/Documents/UV/courses_and_subjects/final_year_proyect/project/data/raw/genomes/archaea/genomic_fna/acidianus_ambivalens/GCF_009428885.1_ASM942888v1_genomic.fna
fna_counter: 1
time_lapse_1: 0.0010306835174560547
time_lapse_2: 0.00875544548034668
mfa.cgr_powers(): power: 13
mfa.cgr_powers(): cumulative: True
mfa.cgr_powers(): m_size: 8192
time_lapse_3: 41.5008339881897
gc_content: 0.3436646411166597
Dq_vals: [1.59665169 1.5979595  1.5994017  1.60100015 1.60278164 1.60477937
 1.60703482 1.60960043 1.61254318 1.61594954 1.61993224 1.62463933
 1.63026539 1.63706238 1.64533832 1.65539988 1.66729993 1.68013853
 1.69123739 1.69669055 1.69319342 1.65580953 1.59296752 1.49697966
 1.4300721  1.38628474 1.35571276 1.33314889 1.31580956 1.30206455
 1.29089337 1.28162692 1.27380972 1.26712149 1.26133049 1.2562648
 1.25179415 1.24781795 1.24425724 1.24104918]
r_squared: [0.83403932 0.83552795 0.83716463 0.83897246 0.8409795  0.84322026
 0.84573754 0.84858501 0.85183078 0.85

### Reciprocal of powers of 2 as epsilon values

fna_counter: 1  
time_lapse_1: 0.0012087821960449219  
time_lapse_2: 0.00923776626586914  
time_lapse_3: 34.21620750427246  
gc_content: 0.3436646411166597  
Dq_vals: [1.90374827 1.90379139 1.90383691 1.9038849  1.90393531 1.90398789  
 1.90404201 1.9040963  1.90414802 1.9041919  1.90421787 1.90420648  
 1.90411954 1.90387974 1.90332357 1.90208683 1.89932641 1.89320285  
 1.88071936 1.85859108 1.82542535 1.74107582 1.68476739 1.61289989  
 1.55711319 1.5182496  1.49008051 1.46876135 1.45208798 1.43869691  
 1.42769953 1.418497   1.41067393 1.40393428 1.39806163 1.39289416  
 1.38830839 1.38420848 1.38051882 1.37717899]  
r_squared: [0.89373615 0.89478934 0.89594759 0.89722735 0.89864861 0.90023596  
 0.9020199  0.9040387  0.90634092 0.9089889  0.91206366 0.91567159  
 0.91995367 0.92509695 0.93134559 0.93899803 0.94833826 0.95934435  
 0.97107779 0.98195411 0.99035558 0.19724959 0.99734307 0.99664264  
 0.99224842 0.98785396 0.98477619 0.98269512 0.98123323 0.98015503  
 0.97932775 0.97867419 0.97814644 0.97771299 0.97735211 0.97704828  
 0.97679007 0.97656884 0.97637795 0.9762122  0.97606744]  
Dq: 0.5270388819557263  
TGAAC  

In [26]:
print("default_epsilon_range:")
default_epsilon_range = np.linspace(0, 0.1, 15)
print(default_epsilon_range)
print("default_q_range:") 
print(np.linspace(-20, 20, 41))

print(f"default_epsilon_range[1]: {default_epsilon_range[1]}")
print(f"1 / default_epsilon_range[1]: {1 / default_epsilon_range[1]}")
print(f"np.log(1 / default_epsilon_range[1]) / np.log(2) = {np.log(1 / default_epsilon_range[1]) / np.log(2)}")

default_epsilon_range:
[0.         0.00714286 0.01428571 0.02142857 0.02857143 0.03571429
 0.04285714 0.05       0.05714286 0.06428571 0.07142857 0.07857143
 0.08571429 0.09285714 0.1       ]
default_q_range:
[-20. -19. -18. -17. -16. -15. -14. -13. -12. -11. -10.  -9.  -8.  -7.
  -6.  -5.  -4.  -3.  -2.  -1.   0.   1.   2.   3.   4.   5.   6.   7.
   8.   9.  10.  11.  12.  13.  14.  15.  16.  17.  18.  19.  20.]
default_epsilon_range[1]: 0.0071428571428571435
1 / default_epsilon_range[1]: 140.0
np.log(1 / default_epsilon_range[1]) / np.log(2) = 7.129283016944966


### Arbitrary epsilon values

fna_counter: 1  
time_lapse_1: 0.0009219646453857422  
time_lapse_2: 0.006658792495727539  
time_lapse_3: 127.31039357185364  
gc_content: 0.3436646411166597  

Dq_vals: [2.9724008  2.96608834 2.95907403 2.9512325  2.94240688 2.93239852  
 2.92095256 2.90773737 2.89231464 2.874095   2.85227066 2.8257109  
 2.7927965  2.75115384 2.69723035 2.62565655 2.52858811 2.39757491  
 2.24130123 2.10360574 1.99997617 1.85832985 1.80612768 1.76401436  
 1.73139942 1.70643349 1.68710398 1.67182474 1.65947597 1.64928964  
 1.64073677 1.63344658 1.62715338 1.62166236 1.61682739 1.61253662  
 1.6087028  1.60525671 1.6021426  1.59931501]  
r_squared: [0.997559   0.99755421 0.99754904 0.99754347 0.99753745 0.99753097  
 0.997524   0.99751652 0.99750851 0.99749993 0.99749082 0.99748147  
 0.99747324 0.99747118 0.99749215 0.99758722 0.99788552 0.99858063  
 0.9994462  0.99984955 1.         0.04118279 0.99970699 0.99973483  
 0.99978949 0.99981242 0.99981635 0.99981392 0.9998099  0.99980563  
 0.99980138 0.9997972  0.99979304 0.99978889 0.99978475 0.9997806  
 0.99977646 0.99977232 0.99976819 0.99976408 0.99976   ]  
Dq: 1.3730857893757518  
TGAAC  

In [27]:
Dq_vals_0_string= """2.9724008  2.96608834 2.95907403 2.9512325  2.94240688 2.93239852
 2.92095256 2.90773737 2.89231464 2.874095   2.85227066 2.8257109
 2.7927965  2.75115384 2.69723035 2.62565655 2.52858811 2.39757491
 2.24130123 2.10360574 1.99997617 1.85832985 1.80612768 1.76401436
 1.73139942 1.70643349 1.68710398 1.67182474 1.65947597 1.64928964
 1.64073677 1.63344658 1.62715338 1.62166236 1.61682739 1.61253662
 1.6087028  1.60525671 1.6021426  1.59931501"""
r_squared_0_string= """0.997559   0.99755421 0.99754904 0.99754347 0.99753745 0.99753097
 0.997524   0.99751652 0.99750851 0.99749993 0.99749082 0.99748147
 0.99747324 0.99747118 0.99749215 0.99758722 0.99788552 0.99858063
 0.9994462  0.99984955 1.         0.04118279 0.99970699 0.99973483
 0.99978949 0.99981242 0.99981635 0.99981392 0.9998099  0.99980563
 0.99980138 0.9997972  0.99979304 0.99978889 0.99978475 0.9997806
 0.99977646 0.99977232 0.99976819 0.99976408 0.99976   """
Dq_vals_0 = np.fromstring(Dq_vals_0_string, sep=' ')
print(Dq_vals_0)

r_squared_0 = np.fromstring(r_squared_0_string, sep=' ')
print(r_squared_0)


[2.9724008  2.96608834 2.95907403 2.9512325  2.94240688 2.93239852
 2.92095256 2.90773737 2.89231464 2.874095   2.85227066 2.8257109
 2.7927965  2.75115384 2.69723035 2.62565655 2.52858811 2.39757491
 2.24130123 2.10360574 1.99997617 1.85832985 1.80612768 1.76401436
 1.73139942 1.70643349 1.68710398 1.67182474 1.65947597 1.64928964
 1.64073677 1.63344658 1.62715338 1.62166236 1.61682739 1.61253662
 1.6087028  1.60525671 1.6021426  1.59931501]
[0.997559   0.99755421 0.99754904 0.99754347 0.99753745 0.99753097
 0.997524   0.99751652 0.99750851 0.99749993 0.99749082 0.99748147
 0.99747324 0.99747118 0.99749215 0.99758722 0.99788552 0.99858063
 0.9994462  0.99984955 1.         0.04118279 0.99970699 0.99973483
 0.99978949 0.99981242 0.99981635 0.99981392 0.9998099  0.99980563
 0.99980138 0.9997972  0.99979304 0.99978889 0.99978475 0.9997806
 0.99977646 0.99977232 0.99976819 0.99976408 0.99976   ]


In [28]:
r_squared_0

array([0.997559  , 0.99755421, 0.99754904, 0.99754347, 0.99753745,
       0.99753097, 0.997524  , 0.99751652, 0.99750851, 0.99749993,
       0.99749082, 0.99748147, 0.99747324, 0.99747118, 0.99749215,
       0.99758722, 0.99788552, 0.99858063, 0.9994462 , 0.99984955,
       1.        , 0.04118279, 0.99970699, 0.99973483, 0.99978949,
       0.99981242, 0.99981635, 0.99981392, 0.9998099 , 0.99980563,
       0.99980138, 0.9997972 , 0.99979304, 0.99978889, 0.99978475,
       0.9997806 , 0.99977646, 0.99977232, 0.99976819, 0.99976408,
       0.99976   ])

# Comparison
epsilon values:  1/(2^(13)), 1/(2^(12)), ..., 1/2

## Arbitrary epsilon values with powers of 2 as values

In [29]:
# measure time
initial_time = time.time()
    
for dir in directory_paths[:1]:
    dir_files = dir.glob('*')
    fna_counter = 0
    seq = ""
    genome_file = None
    for file in dir_files:
        if file.name.endswith('.fna'):
            fna_counter += 1
            genome_file = file
            print(f"genome_file: {genome_file}")
    print(f"fna_counter: {fna_counter}")
    assert( fna_counter == 1 or fna_counter == 0 )
    time_stamp_1 = time.time()
    time_lapse_1 = time_stamp_1 - initial_time
    print(f"time_lapse_1: {time_lapse_1}")
    seq, metadata = mfa_processor.extract_sequence( genome_file )
    time_stamp_2 = time.time()
    time_lapse_2 = time_stamp_2 - time_stamp_1
    print(f"time_lapse_2: {time_lapse_2}")
    epsilon_values_power_13 = [1 / 2**i for i in range(13, 0, -1)]
    gc_content, Dq_vals, r_squared, Dq = mfa_processor.compute_gc_Dq(
        seq, epsilon_range = epsilon_values_power_13, use_powers = False)
    time_stamp_3 = time.time()
    time_lapse_3 = time_stamp_3 - time_stamp_2
    print(f"time_lapse_3: {time_lapse_3}")
    
    print(f"gc_content: {gc_content}")
    print(f"Dq_vals: {Dq_vals}")
    print(f"r_squared: {r_squared}")
    print(f"Dq: {Dq}")

    if seq is not None:
        print(seq[-5:])

genome_file: /home/diegoflm/Documents/UV/courses_and_subjects/final_year_proyect/project/data/raw/genomes/archaea/genomic_fna/acidianus_ambivalens/GCF_009428885.1_ASM942888v1_genomic.fna
fna_counter: 1
time_lapse_1: 0.002160787582397461
time_lapse_2: 0.007711887359619141
time_lapse_3: 130.30656003952026
gc_content: 0.3436646411166597
Dq_vals: [1.59665169 1.5979595  1.5994017  1.60100015 1.60278164 1.60477937
 1.60703482 1.60960043 1.61254318 1.61594954 1.61993224 1.62463933
 1.63026539 1.63706238 1.64533832 1.65539988 1.66729993 1.68013853
 1.69123739 1.69669055 1.69319342 1.65580953 1.59296752 1.49697966
 1.4300721  1.38628474 1.35571276 1.33314889 1.31580956 1.30206455
 1.29089337 1.28162692 1.27380972 1.26712149 1.26133049 1.2562648
 1.25179415 1.24781795 1.24425724 1.24104918]
r_squared: [0.83403932 0.83552795 0.83716463 0.83897246 0.8409795  0.84322026
 0.84573754 0.84858501 0.85183078 0.85556235 0.85989346 0.86497385
 0.87100255 0.8782453  0.88705359 0.89786991 0.91115656 0.92704

In [30]:
epsilon_values_power_13 = [1 / 2**i for i in range(13, 0, -1)]
epsilon_values_power_13

[0.0001220703125,
 0.000244140625,
 0.00048828125,
 0.0009765625,
 0.001953125,
 0.00390625,
 0.0078125,
 0.015625,
 0.03125,
 0.0625,
 0.125,
 0.25,
 0.5]

## Powers

In [31]:
# measure time
initial_time = time.time()
    
for dir in directory_paths[:1]:
    dir_files = dir.glob('*')
    fna_counter = 0
    seq = ""
    genome_file = None
    for file in dir_files:
        if file.name.endswith('.fna'):
            fna_counter += 1
            genome_file = file
            print(f"genome_file: {genome_file}")
    print(f"fna_counter: {fna_counter}")
    assert( fna_counter == 1 or fna_counter == 0 )
    time_stamp_1 = time.time()
    time_lapse_1 = time_stamp_1 - initial_time
    print(f"time_lapse_1: {time_lapse_1}")
    seq, metadata = mfa_processor.extract_sequence( genome_file )
    time_stamp_2 = time.time()
    time_lapse_2 = time_stamp_2 - time_stamp_1
    print(f"time_lapse_2: {time_lapse_2}")
    gc_content, Dq_vals, r_squared, Dq = mfa_processor.compute_gc_Dq(seq, use_powers = True, power = 13)
    time_stamp_3 = time.time()
    time_lapse_3 = time_stamp_3 - time_stamp_2
    print(f"time_lapse_3: {time_lapse_3}")
    
    print(f"gc_content: {gc_content}")
    print(f"Dq_vals: {Dq_vals}")
    print(f"r_squared: {r_squared}")
    print(f"Dq: {Dq}")

    if seq is not None:
        print(seq[-5:])

genome_file: /home/diegoflm/Documents/UV/courses_and_subjects/final_year_proyect/project/data/raw/genomes/archaea/genomic_fna/acidianus_ambivalens/GCF_009428885.1_ASM942888v1_genomic.fna
fna_counter: 1
time_lapse_1: 0.00031447410583496094
time_lapse_2: 0.005953550338745117
mfa.cgr_powers(): power: 13
mfa.cgr_powers(): cumulative: True
mfa.cgr_powers(): m_size: 8192
time_lapse_3: 39.915374755859375
gc_content: 0.3436646411166597
Dq_vals: [1.59665169 1.5979595  1.5994017  1.60100015 1.60278164 1.60477937
 1.60703482 1.60960043 1.61254318 1.61594954 1.61993224 1.62463933
 1.63026539 1.63706238 1.64533832 1.65539988 1.66729993 1.68013853
 1.69123739 1.69669055 1.69319342 1.65580953 1.59296752 1.49697966
 1.4300721  1.38628474 1.35571276 1.33314889 1.31580956 1.30206455
 1.29089337 1.28162692 1.27380972 1.26712149 1.26133049 1.2562648
 1.25179415 1.24781795 1.24425724 1.24104918]
r_squared: [0.83403932 0.83552795 0.83716463 0.83897246 0.8409795  0.84322026
 0.84573754 0.84858501 0.85183078 

### arbitrary epsilon:  
time_lapse_3: 133.3210847377777
gc_content: 0.3436646411166597
Dq_vals: [1.59665169 1.5979595  1.5994017  1.60100015 1.60278164 1.60477937
 1.60703482 1.60960043 1.61254318 1.61594954 1.61993224 1.62463933
 1.63026539 1.63706238 1.64533832 1.65539988 1.66729993 1.68013853
 1.69123739 1.69669055 1.69319342 1.65580953 1.59296752 1.49697966
 1.4300721  1.38628474 1.35571276 1.33314889 1.31580956 1.30206455
 1.29089337 1.28162692 1.27380972 1.26712149 1.26133049 1.2562648
 1.25179415 1.24781795 1.24425724 1.24104918]
r_squared: [0.83403932 0.83552795 0.83716463 0.83897246 0.8409795  0.84322026
 0.84573754 0.84858501 0.85183078 0.85556235 0.85989346 0.86497385
 0.87100255 0.8782453  0.88705359 0.89786991 0.91115656 0.92704184
 0.94450048 0.96158061 0.97598996 0.00112388 0.99078319 0.98828906
 0.97737406 0.96894366 0.96363211 0.96023898 0.9579623  0.95635033
 0.95515849 0.95424904 0.95353899 0.95297509 0.95252139 0.95215265
 0.95185058 0.95160159 0.95139535 0.95122388 0.9510809 ]
Dq: 0.45564137739165345
TGAAC

### Powers of 2:
time_lapse_3: 51.37131690979004
gc_content: 0.3436646411166597
Dq_vals: [1.59665169 1.5979595  1.5994017  1.60100015 1.60278164 1.60477937
 1.60703482 1.60960043 1.61254318 1.61594954 1.61993224 1.62463933
 1.63026539 1.63706238 1.64533832 1.65539988 1.66729993 1.68013853
 1.69123739 1.69669055 1.69319342 1.65580953 1.59296752 1.49697966
 1.4300721  1.38628474 1.35571276 1.33314889 1.31580956 1.30206455
 1.29089337 1.28162692 1.27380972 1.26712149 1.26133049 1.2562648
 1.25179415 1.24781795 1.24425724 1.24104918]
r_squared: [0.83403932 0.83552795 0.83716463 0.83897246 0.8409795  0.84322026
 0.84573754 0.84858501 0.85183078 0.85556235 0.85989346 0.86497385
 0.87100255 0.8782453  0.88705359 0.89786991 0.91115656 0.92704184
 0.94450048 0.96158061 0.97598996 0.00112388 0.99078319 0.98828906
 0.97737406 0.96894366 0.96363211 0.96023898 0.9579623  0.95635033
 0.95515849 0.95424904 0.95353899 0.95297509 0.95252139 0.95215265
 0.95185058 0.95160159 0.95139535 0.95122388 0.9510809 ]
Dq: 0.45564137739165345
TGAAC

In [46]:
Dq_vals_arbitrary_epsilon = np.fromstring("""
    1.59665169 1.5979595  1.5994017  1.60100015 1.60278164 1.60477937
    1.60703482 1.60960043 1.61254318 1.61594954 1.61993224 1.62463933
    1.63026539 1.63706238 1.64533832 1.65539988 1.66729993 1.68013853
    1.69123739 1.69669055 1.69319342 1.65580953 1.59296752 1.49697966
    1.4300721  1.38628474 1.35571276 1.33314889 1.31580956 1.30206455
    1.29089337 1.28162692 1.27380972 1.26712149 1.26133049 1.2562648
    1.25179415 1.24781795 1.24425724 1.24104918
    """, sep=' ')
r_squared_arbitrary_epsilon = np.fromstring("""
    0.83403932 0.83552795 0.83716463 0.83897246 0.8409795  0.84322026
    0.84573754 0.84858501 0.85183078 0.85556235 0.85989346 0.86497385
    0.87100255 0.8782453  0.88705359 0.89786991 0.91115656 0.92704184
    0.94450048 0.96158061 0.97598996 0.00112388 0.99078319 0.98828906
    0.97737406 0.96894366 0.96363211 0.96023898 0.9579623  0.95635033
    0.95515849 0.95424904 0.95353899 0.95297509 0.95252139 0.95215265
    0.95185058 0.95160159 0.95139535 0.95122388 0.9510809
""", sep=' ')
print(r_squared_arbitrary_epsilon)

[0.83403932 0.83552795 0.83716463 0.83897246 0.8409795  0.84322026
 0.84573754 0.84858501 0.85183078 0.85556235 0.85989346 0.86497385
 0.87100255 0.8782453  0.88705359 0.89786991 0.91115656 0.92704184
 0.94450048 0.96158061 0.97598996 0.00112388 0.99078319 0.98828906
 0.97737406 0.96894366 0.96363211 0.96023898 0.9579623  0.95635033
 0.95515849 0.95424904 0.95353899 0.95297509 0.95252139 0.95215265
 0.95185058 0.95160159 0.95139535 0.95122388 0.9510809 ]
